File: backend/app/tuning/store.py
"""Persistent per-craft tune-iteration history.

The product's whole philosophy is iterative tuning across multiple flights
(see the UX spec: "Session: Chimera7 / Tune #1 - Baseline / Tune #2 -
Applied / ..."), which means iteration history must survive backend
restarts -- unlike the in-memory Blackbox session store in api/routes.py
(which is fine to lose on restart, since a session is just "the log you're
looking at right now"). A single-user kiosk app on one Pi has no need for a
real database; one JSON file per craft is simple, human-inspectable, and
sufficient.
"""
from __future__ import annotations

import json
import re
import time
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Optional

from app import config


@dataclass
class Iteration:
    number: int                      # 1-based, sequential per craft
    timestamp: float                 # seconds since epoch (time.time())
    label: str                       # "Baseline" | "Applied" | "Current" -- display only
    applied_changes: list = field(default_factory=list)   # [{"parameter":..., "from":..., "to":...}]
    analysis_summary: dict = field(default_factory=dict)  # snapshot of GET /api/analysis/summary


def craft_id_from_name(craft_name: Optional[str]) -> str:
    """Sanitize a craft name into a filesystem-safe id. Unnamed/unknown
    craft still needs a stable id to accumulate history against, rather
    than silently dropping iteration tracking -- "unnamed" is a deliberate,
    stable fallback bucket, not a random/per-session id, so history still
    accumulates across sessions for a craft that never got a name set."""
    if not craft_name or not craft_name.strip():
        return "unnamed"
    slug = re.sub(r"[^a-zA-Z0-9_-]+", "_", craft_name.strip().lower())
    return slug or "unnamed"


def _store_path(craft_id: str) -> Path:
    return config.TUNING_STORE_DIR / f"{craft_id}.json"


def load_iterations(craft_id: str) -> list[Iteration]:
    path = _store_path(craft_id)
    if not path.exists():
        return []
    with open(path) as f:
        raw = json.load(f)
    return [Iteration(**item) for item in raw]


def save_iteration(craft_id: str, label: str, applied_changes: list, analysis_summary: dict) -> Iteration:
    """Append a new iteration for this craft and persist it. Returns the
    saved Iteration (with its assigned sequential number)."""
    existing = load_iterations(craft_id)
    number = (existing[-1].number + 1) if existing else 1
    iteration = Iteration(
        number=number,
        timestamp=time.time(),
        label=label,
        applied_changes=applied_changes,
        analysis_summary=analysis_summary,
    )
    existing.append(iteration)
    path = _store_path(craft_id)
    with open(path, "w") as f:
        json.dump([asdict(it) for it in existing], f, indent=2)
    return iteration


def get_latest_iteration(craft_id: str) -> Optional[Iteration]:
    iterations = load_iterations(craft_id)
    return iterations[-1] if iterations else None

File: mobile/src/tuning/store.ts
/** Persistent per-craft tune-iteration history.
 *
 * Ported from backend/app/tuning/store.py, adapted from "one JSON file per
 * craft" (fine for a single-user kiosk Pi) to a local SQLite database via
 * expo-sqlite, since a mobile app doesn't have the Python version's
 * filesystem-as-database convenience but does get a real embedded SQL
 * engine for free. The product's whole philosophy is iterative tuning
 * across multiple flights, so this history must survive app restarts.
 */
import { openDatabaseAsync, type SQLiteDatabase } from 'expo-sqlite';
import type { AnalysisSummary } from './analysisSummary';

export interface AppliedChange {
  parameter: string;
  from: number | null;
  to: number | null;
}

export interface Iteration {
  number: number; // 1-based, sequential per craft
  timestamp: number; // seconds since epoch
  label: string; // "Baseline" | "Applied" | "Current" -- display only
  appliedChanges: AppliedChange[];
  analysisSummary: AnalysisSummary;
}

let dbPromise: Promise<SQLiteDatabase> | null = null;

function getDb(): Promise<SQLiteDatabase> {
  if (!dbPromise) {
    dbPromise = openDatabaseAsync('tuning_history.db').then(async (db) => {
      await db.execAsync(`
        CREATE TABLE IF NOT EXISTS iterations (
          craft_id TEXT NOT NULL,
          number INTEGER NOT NULL,
          timestamp REAL NOT NULL,
          label TEXT NOT NULL,
          applied_changes TEXT NOT NULL,
          analysis_summary TEXT NOT NULL,
          PRIMARY KEY (craft_id, number)
        );
      `);
      return db;
    });
  }
  return dbPromise;
}

/** Sanitize a craft name into a storage-safe id. Unnamed/unknown craft
 * still needs a stable id to accumulate history against, rather than
 * silently dropping iteration tracking -- "unnamed" is a deliberate,
 * stable fallback bucket, not a random/per-session id, so history still
 * accumulates across sessions for a craft that never got a name set. */
export function craftIdFromName(craftName: string | null | undefined): string {
  if (!craftName || !craftName.trim()) return 'unnamed';
  const slug = craftName
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9_-]+/g, '_');
  return slug || 'unnamed';
}

interface IterationRow {
  number: number;
  timestamp: number;
  label: string;
  applied_changes: string;
  analysis_summary: string;
}

function rowToIteration(row: IterationRow): Iteration {
  return {
    number: row.number,
    timestamp: row.timestamp,
    label: row.label,
    appliedChanges: JSON.parse(row.applied_changes),
    analysisSummary: JSON.parse(row.analysis_summary),
  };
}

export async function loadIterations(craftId: string): Promise<Iteration[]> {
  const db = await getDb();
  const rows = await db.getAllAsync<IterationRow>(
    'SELECT number, timestamp, label, applied_changes, analysis_summary FROM iterations WHERE craft_id = ? ORDER BY number ASC',
    [craftId]
  );
  return rows.map(rowToIteration);
}

/** Append a new iteration for this craft and persist it. Returns the saved
 * Iteration (with its assigned sequential number). */
export async function saveIteration(
  craftId: string,
  label: string,
  appliedChanges: AppliedChange[],
  analysisSummary: AnalysisSummary
): Promise<Iteration> {
  const existing = await loadIterations(craftId);
  const number = existing.length > 0 ? existing[existing.length - 1].number + 1 : 1;
  const timestamp = Date.now() / 1000;

  const db = await getDb();
  await db.runAsync(
    'INSERT INTO iterations (craft_id, number, timestamp, label, applied_changes, analysis_summary) VALUES (?, ?, ?, ?, ?, ?)',
    [craftId, number, timestamp, label, JSON.stringify(appliedChanges), JSON.stringify(analysisSummary)]
  );

  return { number, timestamp, label, appliedChanges, analysisSummary };
}

export async function getLatestIteration(craftId: string): Promise<Iteration | null> {
  const iterations = await loadIterations(craftId);
  return iterations.length > 0 ? iterations[iterations.length - 1] : null;
}
